### ISSUE TITLE
`event.set_results` with `stdout` as key gives wrong action results

### ISSUE THREAD

**Issue by sed-i** [canonical/operator#689]
## Intro
After an action completes, I am using `event.set_results` with key `"stdout"`.
For full context see [here](https://github.com/canonical/cos-configuration-k8s-operator/blob/3e9e1c8373060542672b6f47849c2b4313a547bc/src/charm.py#L233).

`event.set_results` with `stdout` as key prints out odd results.
I.e. `juju run-action` results do not match what is set by `event.set_results`.
Maybe this is a juju issue and not ops.

## Environment

| Tool | Version |
|------|---------|
|  juju    |   2.9.24      |
| ops | [0f752b8](https://github.com/canonical/operator/commit/0f752b8592e76442284e4ef615575a1e744cd041) |

### Case 1
```python
event.set_results({"stdout": "", "stdout-key": "", "and-something": "else"})
```

### Case 2
```python
event.set_results({"stdout": "stdout-non-empty-value", "stdout-key": "", "and-something": "else"})
```

### Case 1
```
    stdout: ""
    stdout-key: ""
    and-something: else
```

### Case 2
```
    stdout: "stdout-non-empty-value"
    stdout-key: ""
    and-something: else
```

### Case 1
The `"stdout": ""` key-val pair is dropped from the results:
```
    stdout-key: ""
    and-something: else
```

### Case 2
The `stdout` key got capitalized:
```
    Stdout: "stdout-non-empty-value"
    stdout-key: ""
    and-something: else
```

### Case 1
```
$ juju run-action cos-config/0 sync-now --wait
unit-cos-config-0:
  UnitId: cos-config/0
  id: "18"
  log:
  - 2022-02-10 22:33:56 +0000 UTC Calling git-sync with --one-time...
  - '2022-02-10 22:33:57 +0000 UTC Warning: I0210 22:33:56.268703      30 main.go:454]
    "level"=0 "msg"="starting up" "pid"=30 "args"=["/git-sync","--repo","https://github.com/canonical/avalanche-k8s-operator","--branch","feature/fix_dashboards","--rev","HEAD","--depth","1","--root","/git","--dest","repo","--one-time"]'
  results:
    and-something: else
    stdout-key: ""
  status: completed
  timing:
    completed: 2022-02-10 22:33:57 +0000 UTC
    enqueued: 2022-02-10 22:33:54 +0000 UTC
    started: 2022-02-10 22:33:56 +0000 UTC
```

### Case 2
```
$ juju run-action cos-config/0 sync-now --wait
unit-cos-config-0:
  UnitId: cos-config/0
  id: "20"
  log:
  - 2022-02-10 22:39:30 +0000 UTC Calling git-sync with --one-time...
  - '2022-02-10 22:39:30 +0000 UTC Warning: I0210 22:39:29.990093      28 main.go:454]
    "level"=0 "msg"="starting up" "pid"=28 "args"=["/git-sync","--repo","https://github.com/canonical/avalanche-k8s-operator","--branch","feature/fix_dashboards","--rev","HEAD","--depth","1","--root","/git","--dest","repo","--one-time"]'
  results:
    Stdout: stdout-non-empty-value
    and-something: else
    stdout-key: ""
  status: completed
  timing:
    completed: 2022-02-10 22:39:31 +0000 UTC
    enqueued: 2022-02-10 22:39:29 +0000 UTC
    started: 2022-02-10 22:39:29 +0000 UTC
```

**Comment by rwcarlsen** [comment]
I dug through ops code for a while - and also looked at juju.  This is all on the juju side.  Specifically `cmd/juju/commands/run.go:191` is the "Stdout" param (it's intentional).  And reguarding the dropped empty keys - the ops code, the juju action-set, and juju actions api code preserver empty valued keys.   `github.com/juju/cmd` is where the results formatting is done - I suspect that the yaml-formatting code there is responsible for dropping the empty values.

**Comment by sed-i** [comment]
Thanks for looking into this!

> Specifically cmd/juju/commands/run.go:191 is the "Stdout" param (it's intentional)

- Should ops document and prevent using "Stdout" and "stdout" as keys?
- What would good practice for forwarding stdout from container.exec to action results? Currently I use:

```python
process = self.container.exec(...)
stdout, warnings = process.wait_output()
event.set_results({"stdout": stdout})
```

> I suspect that the yaml-formatting code there is responsible for dropping the empty values

In the cases above, the empty value is dropped only if the key is "stdout", otherwise it is not dropped.

**Comment by sed-i** [comment]
> What is your juju controller version?

2.9.24

**Comment by rwcarlsen** [comment]
An even worse outcome is that if you `print('foo')` inside your action *and* set the `stdout` param in results, your `stdout` results param gets clobbered by the actual action stdout.  We should probably put in a guard that yells loud if someone tries to set the `stdout` key in the results dict.  But I'd still like to understand better why this is happening.  Sounds like a Monday job.

**Comment by rwcarlsen** [comment]
Actually it appears to be non-deterministic.  Sometimes the "stdout" param set in the event results overrides the actual stdout (from e.g. print) and sometimes actual stdout wins and overwrites the param val.

**Comment by rwcarlsen** [comment]
This was fixed in juju/juju#13748.
